Feature/dome

petersteneteg opened the pull request:
Enables clustered dome rendering with sgct

petersteneteg committed:
SGCT: basic support

SGCT: vcpkg fixes

SGCT: WIP

SGCT: wip

petersteneteg committed:
SGCT: updates and fixes

petersteneteg committed:
SGCT: inviwo gui server support

petersteneteg committed:
SGCT: new sgct version

petersteneteg committed:
SGCT: workspacewatcher fs update

ResearchDaniel commented on apps/inviwodome/CMakeLists.txt:
```diff
@@ -0,0 +1,99 @@
+    
+target_link_libraries(inviwo-dome 
+    PRIVATE
+        Tracy::TracyClient
```
Tracy should not be required?

petersteneteg replied:
Might remove that in the future, when we are done optimizing. This is still work in progress.

petersteneteg replied:
Is optional now

ResearchDaniel replied:
In another commit?

petersteneteg replied:
now it's also pushed

ResearchDaniel commented on apps/inviwodome/CMakeLists.txt:
```diff
@@ -0,0 +1,99 @@
+endif()
+
+
+file(GENERATE 
```
Add commment about what this does

ResearchDaniel commented on apps/inviwodome/DomePreset.json:
```diff
@@ -0,0 +1,66 @@
+      "hidden" : true,
+
+      "cacheVariables": {
+        // We use a static build
```
Ok, but please comment why we need to use a static build?

petersteneteg replied:
We don't but it makes it easier to profile with tracy

ResearchDaniel commented on apps/inviwodome/DomePreset.json:
```diff
@@ -0,0 +1,66 @@
+        "IVW_MODULE_OPENGLQT":        { "type": "BOOL", "value": "OFF" },
+
+
+        // OpenCL Needs to use glfw
```
We could disable OpenCL in case it requires extra “work”.

petersteneteg replied:
No extra work apart from settig the openglSupplier

ResearchDaniel commented on apps/inviwodome/DomePreset.json:
```diff
@@ -0,0 +1,66 @@
+        "IVW_MODULE_GLFW":            { "type": "BOOL", "value": "ON" },
+        "IVW_MODULE_OPENCL_OpenGLSupplier": "InviwoGLFWModule",
+
+        // Disauble other target not needed for doem
```
```suggestion
        // Disable targets not needed for the dome application.
```

ResearchDaniel commented on apps/inviwodome/include/inviwo/dome/sgctmanager.h:
```diff
@@ -0,0 +1,85 @@
+class CameraProperty;
+class Shader;
+
+struct SgctManager : public ProcessorNetworkObserver {
```
Nitpick
```suggestion
struct SGCTManager : public ProcessorNetworkObserver {
```

ResearchDaniel commented on apps/inviwodome/include/inviwo/dome/sgctmanager.h:
```diff
@@ -0,0 +1,85 @@
+    CanvasProcessor* getCanvas() { return canvases.empty() ? nullptr : canvases.front(); }
+    CameraProperty* getCamera() { return cameras.empty() ? nullptr : cameras.front(); }
+
+    void setupUpInteraction(GLFWwindow* win);
```
```suggestion
    void setupInteraction(GLFWwindow* win);
```

ResearchDaniel commented on apps/inviwodome/include/inviwo/dome/sgctmanager.h:
```diff
@@ -0,0 +1,85 @@
+    void evaluate(const ::sgct::RenderData& renderData);
+
+    void createShader();
+    void copy();
```
This does not make much sense. What is copied?

I assume that createShader and copy are just used internally. Could make them private

ResearchDaniel commented on apps/inviwodome/packaging.cmake:
```diff
@@ -0,0 +1,109 @@
+endif()
+
+install(DIRECTORY ${IVW_ROOT_DIR}/data/  DESTINATION ${IVW_RESOURCE_INSTALL_PREFIX}data  COMPONENT Datasets)
+#install(DIRECTORY ${IVW_ROOT_DIR}/tests/ DESTINATION ${IVW_RESOURCE_INSTALL_PREFIX}tests COMPONENT Testing)
```
Remove

ResearchDaniel commented on apps/inviwodome/scripts/2_nodes.xml:
```diff
@@ -0,0 +1,39 @@
+    </Window>
+  </Node>
+  <Node address="127.0.0.2" port="20402">
+  <!--Node address="10.245.163.128" port="20402"-->
```
```suggestion
```

ResearchDaniel commented on apps/inviwodome/scripts/martin.xml:
Having a file called Martin does not make too much sense :)

ResearchDaniel commented on apps/inviwodome/scripts/1_nodes.xml:
Could we either add a readme file or add a comment in each config file about what they are configuring?
Unclear what 1 and 2 in x_nodes are referring to

petersteneteg replied:
it uses 1 or 2 nodes in a cluster..

ResearchDaniel commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,358 @@
+#include <inviwo/tracy/tracyopengl.h>
+
+/*
+void* operator new(size_t count) {
```
Remove

petersteneteg replied:
Added a comment

ResearchDaniel replied:
Wrap around a TRACY_DEFINED?

ResearchDaniel commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,358 @@
+        // Initialize all modules
+        app.registerModules(inviwo::getModuleList());
+
+        // app.getModuleByType<inviwo::GLFWModule>()->setWaitForOpenGL(false);
```
Remove
```suggestion
```

ResearchDaniel commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,358 @@
+    std::mutex commandsMutex;
+    std::vector<inviwo::SgctCommand> commands;
+
+    void setCamerasToSgct(inviwo::ProcessorNetwork& net) {
```
```suggestion
    void setCamerasToSGCT(inviwo::ProcessorNetwork& net) {
```

ResearchDaniel commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,358 @@
+
+        if (sgct::Engine::instance().isMaster()) {
+            app.getWorkspaceManager()->load(
+                app.getPath(inviwo::PathType::Workspaces, "/boron.inv"));
```
No need for hardcoded workspace?

petersteneteg replied:
why not, it is nice to have a default.

ResearchDaniel replied:
Ok, but otherwise one would see the getting started view, which might be better?

petersteneteg replied:
There is no QT app here so you will see nothing

ResearchDaniel commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,358 @@
+        TRACY_ZONE_SCOPED_NC("Draw", 0xAAAA00);
+        TRACY_GPU_ZONE_C("Draw", 0xAAAA00);
+
+        // Save State
```
I don’t understand this. Why is it necessary to get the draw buffers each draw?

petersteneteg replied:
Since sgct will break otherwise

ResearchDaniel commented on apps/inviwodome/include/inviwo/dome/sgctmanager.h:
```diff
@@ -0,0 +1,85 @@
+    std::vector<std::unique_ptr<GLFWUserData>> userdata;
+    std::vector<std::unique_ptr<GLFWWindowEventManager>> interactionManagers;
+    std::unique_ptr<Shader> copyShader;
+    std::function<void(bool)> onStatChange;
```
Typo?
```suggestion
    std::function<void(bool)> onStateChange;
```

petersteneteg replied:
Stat used for statistics.

martinfalk replied:
suggestion: `onStatsChange`

ResearchDaniel commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,358 @@
+
+        inviwo::util::OnScopeExit clearNetwork{[&app]() { app.getWorkspaceManager()->clear(); }};
+
+        // Keep the network looked all the time, only unlock in the draw call
```
```suggestion
        // Prevent network evaluation outside the draw call 
```

ResearchDaniel commented on apps/inviwodome/src/sgctmanager.cpp:
```diff
@@ -0,0 +1,238 @@
+
+    if (auto cp = dynamic_cast<CanvasProcessor*>(processor)) {
+        if (canvases.empty()) {
+            eventTarget = cp;
```
Should we log which canvas that was chosen?

petersteneteg replied:
That is usually quite obvious just by looking

ResearchDaniel commented on apps/inviwodome/src/sgctmanager.cpp:
```diff
@@ -0,0 +1,238 @@
+            }
+        },
+        // Todo figure out proper depth.
+        [this](dvec2 p) { return -1.0; }));
```
?

petersteneteg replied:
Interaction is still work in progress.

ResearchDaniel commented on modules/sgct/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+
+target_link_libraries(inviwo-module-sgct PRIVATE sgct::sgct)
+
+# ivw_add_to_module_pack(${CMAKE_CURRENT_SOURCE_DIR}/glsl)
```
Remove

ResearchDaniel commented on modules/sgct/include/inviwo/sgct/datastructures/sgctcamera.h:
```diff
@@ -0,0 +1,79 @@
+
+namespace inviwo {
+
+class IVW_MODULE_SGCT_API SGCTCamera final : public Camera {
```
Could add brief documentation

ResearchDaniel commented on modules/sgct/readme.md:
```diff
@@ -0,0 +1,3 @@
+# sgct Module
+
+Description of the sgct module
```
Update. Would be good with an explanation of what is needed to run in a cluster environment

ResearchDaniel commented on modules/sgct/src/datastructures/sgctcamera.cpp:
```diff
@@ -0,0 +1,182 @@
+
+std::string SGCTCamera::getClassIdentifier() const { return classIdentifier; }
+
+const std::string SGCTCamera::classIdentifier = "SGCTCamera";
```
org.inviwo.SGCTCamera?

petersteneteg replied:
We have not use that for other cameras... Think it shoule be unique enough

ResearchDaniel commented on modules/sgct/src/datastructures/sgctcamera.cpp:
```diff
@@ -0,0 +1,182 @@
+    Camera::updateFrom(source);
+
+    if (auto sc = dynamic_cast<const SGCTCamera*>(&source)) {
+        setFovy(sc->getFovy());
```
Could add comment that setFovy does it’s own modified check

ResearchDaniel commented on modules/sgct/src/datastructures/sgctcamera.cpp:
```diff
@@ -0,0 +1,182 @@
+    mat4 view = glm::make_mat4(renderData.viewMatrix.values);
+    mat4 model = glm::make_mat4(renderData.modelMatrix.values);
+
+    bool modified = true;  // true here to force eval every frame, for testing
```
Should not be true?

ResearchDaniel commented on modules/sgct/src/sgctmodule.cpp:
```diff
@@ -0,0 +1,119 @@
+        };
+
+        try {
+            LogInfoCustom("Dome", "Start Engine");
```
Could replace these logging sources
```suggestion
            LogInfoCustom("SGCT", "Start Engine");
```

ResearchDaniel commented on modules/sgct/src/sgctmodule.cpp:
```diff
@@ -0,0 +1,119 @@
+    std::thread runner_;
+};
+
+// "C:/Users/petst55.AD/Documents/Inviwo/inviwo-dome/apps/inviwodome/scripts/2_nodes.xml"
```
Remove

ResearchDaniel commented on modules/sgct/src/sgctmodule.cpp:
```diff
@@ -0,0 +1,119 @@
+
+SGCTModule::SGCTModule(InviwoApplication* app)
+    : InviwoModule(app, "sgct")
+    , configFileArg_{"",   "config", "The sgct configuration file to use",
```
```suggestion
    , configFileArg_{"",   "config", "Path to the sgct configuration file to use",
```

ResearchDaniel commented on tools/vcpkg/sgct/vcpkg.json:
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "sgct",
```
Should these not be added to the official vcpkg repo instead?

petersteneteg replied:
Sure, ask the sgct developers to add it.

ResearchDaniel commented on vcpkg.json:
```diff
@@ -46,6 +46,54 @@
+        }
+      ]
+    },
+    "dome" : {
```
```suggestion
    "sgct" : {
```

ResearchDaniel commented on vcpkg.json:
```diff
@@ -46,6 +46,54 @@
+      ]
+    },
+    "dome" : {
+      "description": "Inviwo Dome dependencies",
```
```suggestion
      "description": "Inviwo SGCT dependencies",
```

ResearchDaniel reviewed (dismissed):
Inviwo SGCT would be a better name I think

petersteneteg committed:
SGCT: review comments

petersteneteg commented:
I will use inviwo-dome over inviwo-sgct. For two reasons, there is already a inviwo-module-sgct. And no one knows what sgct is. 
If we have other uses of it outside of the dome, we can create more apps in that case. inviwo-vr or what ever.

martinfalk commented on apps/inviwodome/CMakeLists.txt:
```diff
@@ -0,0 +1,91 @@
+    message(STATUS "Package inviwo dome")
+endif()
+
+# Create a helper sciript to start inviwo-dome with multiple instances
```
```suggestion
# Create a helper script to start inviwo-dome with multiple instances
```

martinfalk commented on apps/inviwodome/include/inviwo/dome/workspacewatcher.h:
```diff
@@ -0,0 +1,42 @@
+    std::filesystem::path& path;
+};
+
+}  // namespace inviwo
```
missing new line

martinfalk commented on apps/inviwodome/packaging.cmake:
```diff
@@ -0,0 +1,108 @@
+    if(IVW_PACKAGE_INSTALLER)
+        set(CPACK_GENERATOR "ZIP;NSIS")
+        # The icon to start the application.
+        set(CPACK_NSIS_MUI_ICON "${IVW_ROOT_DIR}\\\\resources\\\\inviwo\\\\inviwo_light.ico")
```
custom icon!

petersteneteg replied:
Im waiting for you to do one :)

martinfalk commented on apps/inviwodome/src/dumpfiles.cpp:
```diff
@@ -0,0 +1,71 @@
+#include <inviwo/dome/dumpfiles.h>
+
+#ifdef WIN32
+#include <Windows.h>
```
```suggestion
#define WIN32_LEAN_AND_MEAN
#include <Windows.h>
```

petersteneteg replied:
we set that globally https://github.com/inviwo/inviwo/blob/69d2062a68c650f78d1f4885f4711059fd6a8db6/cmake/compileoptions.cmake#L204

martinfalk commented on apps/inviwodome/src/dumpfiles.cpp:
```diff
@@ -0,0 +1,71 @@
+
+#endif
+
+}  // namespace inviwo
```
missing new line

martinfalk commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,360 @@
+        inviwo::util::OnScopeExit closeEngine{[]() {
+            LogInfoCustom("Dome", "Stop Engine");
+            // Stop the engine after we clear the inviwo app.
+            // The app will need the glfw to be active to joining
```
```suggestion
            // The app will need the glfw context to be active to join
```

martinfalk commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,360 @@
+        inviwo::util::OnScopeExit clearNetwork{[&app]() { app.getWorkspaceManager()->clear(); }};
+
+        // Keep the network looked to prevent evaluation all the time, only unlock it in the draw
+        // call
```
```suggestion
        // Keep the network locked to prevent evaluating it all the time, only unlock it in the draw
        // call
```

ResearchDaniel replied:
Still a typo in looked

martinfalk commented on apps/inviwodome/src/sgctmanager.cpp:
```diff
@@ -0,0 +1,238 @@
+    }
+    for (auto* camera : processor->getPropertiesByType<CameraProperty>(true)) {
+        if (cameras.empty()) {
+            sgct::Engine::instance().setNearAndFarClippingPlanes(camera->getNearPlaneDist(),
+                                                                 camera->getFarPlaneDist());
```
Could backfire quite easily to only consider the very first camera in the network. Some processors have "dummy" cameras like the orientation indicator.
Or does this assume that all cameras are linked?

martinfalk replied:
ping

petersteneteg replied:
Have not been a issue so far... might be able to traverse the network to find an appropriate camera

martinfalk commented on apps/inviwodome/src/workspacewatcher.cpp:
```diff
@@ -0,0 +1,39 @@
+
+void WorkspaceWatcher::fileChanged(const std::filesystem::path& filename) { path = filename; }
+
+}  // namespace inviwo
```
missing newline

martinfalk commented on modules/sgct/readme.md:
```diff
@@ -0,0 +1,8 @@
+
+Provides a Inviwo Camera (SGCTCamera), that can be handle sgct renderdata 
+
+Also provideds a network facility to sync inviwo network changes between sgct nodes
```
```suggestion
Imports [SGCT](https://github.com/sgct/sgct) into Inviwo and provides an Inviwo Camera (SGCTCamera) that can be handle SGCT renderdata.

Also provideds a network facility to sync Inviwo network changes between SGCT nodes.
```

ResearchDaniel replied:
Still some typos here

martinfalk commented on modules/sgct/src/sgctmodule.cpp:
```diff
@@ -0,0 +1,117 @@
+SGCTModule::SGCTModule(InviwoApplication* app)
+    : InviwoModule(app, "sgct")
+    , configFileArg_{"",    "config", "Path to the sgct configuration file to use",
+                     false, "path",   "sgct json config file"}
```
```suggestion
    , configFileArg_{"",    "config", "Path to the SGCT configuration file to use",
                     false, "path",   "SGCT json config file"}
```

martinfalk reviewed (commented):
some comments and copyright should be updated.

petersteneteg committed:
SGCT: updates, WIP

petersteneteg committed:
SGCT: vcpkg update

petersteneteg committed:
SGCT: Python updates

petersteneteg committed:
SGCT: cmake, make tracy optional

petersteneteg committed:
Vcpkg: baseline update

petersteneteg committed:
SGCT: updates

petersteneteg committed:
SGCT: package fixes

petersteneteg committed:
SGCT: review updates

ResearchDaniel commented on apps/inviwodome/CMakeLists.txt:
```diff
@@ -0,0 +1,96 @@
+set_target_properties(inviwo-dome PROPERTIES 
+    VS_DEBUGGER_WORKING_DIRECTORY $<TARGET_FILE_DIR:inviwo-dome>
+    SGCT_CONFIG_DIR ${CMAKE_CURRENT_LIST_DIR}/scripts
+    TRACY_EXE ${VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/tools/tracy/Tracy.exe
```
Should TRACY_EXE be moved withing IVW_ENABLE_TRACY?

martinfalk commented on apps/inviwodome/DomePreset.json:
```diff
@@ -0,0 +1,98 @@
+      "hidden" : true,
+
+      "cacheVariables": {
+        // Config
```
comments are not part of the JSON standard

petersteneteg replied:
Nopes but cmake is fine with them, so I don't care ;)

martinfalk commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,370 @@
+}
+*/
+
+class Conf {
```
why `Conf` and not `Config` or similar?

martinfalk commented on apps/inviwodome/src/inviwodome.cpp:
```diff
@@ -0,0 +1,370 @@
+    };
+
+    void draw(const sgct::RenderData& renderData) {
+        TRACY_ZONE_SCOPED_NC("Draw", 0xAAAA00);
```
are `TRACY` macros defined even if Tracy is not enabled?

petersteneteg replied:
Yes

martinfalk commented on apps/inviwodome/src/sgctmanager.cpp:
```diff
@@ -0,0 +1,248 @@
+}
+)";
+
+#include <inviwo/core/network/networkvisitor.h>
```
strange location for an include

martinfalk commented on modules/sgct/src/datastructures/sgctcamera.cpp:
```diff
@@ -0,0 +1,194 @@
+
+bool SGCTCamera::equal(const Camera& other) const {
+    if (auto rhs = dynamic_cast<const SGCTCamera*>(&other)) {
+        return equalTo(other) && fovy_ == rhs->fovy_ && extProj_ == rhs->extProj_ &&
```
include `aspectRatio_` set in `setExternal()`?

petersteneteg replied:
the perspective camera doesn't do that

petersteneteg committed:
SGCT: unified build

petersteneteg committed:
SGCT: added sgct settings

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg committed:
SGCT: review fixes

# Conflicts:
#	vcpkg.json

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg committed:
GHA: added debug label for tmate

petersteneteg committed:
GHA: fix

petersteneteg committed:
GHA: Use ccache to speed up building

Msvc is now supported with some care.
Don't check for submodules if vcpkg is used.
Better log pattern matches in gha.
Enable caching of python packages.

petersteneteg committed:
SGCT: clang-tidy fixes

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg committed:
SGCT: clang-tidy fixes

# Conflicts:
#	modules/sgct/include/inviwo/sgct/datastructures/sgctcamera.h

petersteneteg committed:
SGCT: clang-tidy fixes

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg committed:
SGCT: clang-tidy fixes

petersteneteg committed:
GHA: can't update caches, need new keys

petersteneteg committed:
SGCT: format fix
